File: 5. LinkedLists/linked_list_find/linked_list_find.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");

a.next = b;
b.next = c;
c.next = d;

// Iterative
const linkedListFind = (head, target) => {
    while(head !== null){
        if(head === null) return false;
        if(head.val === target) return true;
        head = head.next;
    }
    return false;

}



// Recursive
// const linkedListFind = (head, target) => {
//     if(head === null) return false;
//     if(head.val === target) return true;
//     return linkedListFind(head.next, target);
// }

// Complexity
// Iterative
// n = number of nodes
// Time: O(n)
// Space: O(1)

// Recursive
// n = number of nodes
// Time: O(n)
// Space: O(n)

// let ans = linkedListFind(a, "c");
// console.log(`Ans: ${ans}`);

File: 5. LinkedLists/3. linked_list_find/linked_list_find.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d

# Iterative
def linked_list_find(head, target):
    current = head
    while(current is not None):
        if (current.val == target):
            return True
        current = current.next
    return False

# Recursive
def linked_list_find1(head, target):
    if(head is None):
        return False
    if(head.val == target):
        return True
    return linked_list_find1(head.next, target)


print(f"Answer: {linked_list_find(a, 'c')}") # True
print(f"Answer: {linked_list_find1(a, 'c')}") # True